aten/src/ATen/native/cuda/MultiTensorApply.cuh

```diff
@@ -95,6 +110,70 @@ struct FusedOptimizerTensorListMetadata {
   int start_tensor_this_launch;
 };
 
+bool can_use_static_tensor_list_meta(
+    std::vector<std::vector<at::Tensor>>& tensor_lists,
+    int depth) {
+  const int64_t n_tensors = tensor_lists[0].size();
+  if (n_tensors > depth_to_max_tensors[depth - 1]) {
+    return false;
+  }
+  int64_t num_blocks = 0;
+  for (const auto t : c10::irange(n_tensors)) {
+    const auto numel = tensor_lists[0][t].numel();
+    const auto chunks = numel / kChunkSize + (numel % kChunkSize != 0);
+    num_blocks += chunks;
+    if (num_blocks > depth_to_max_blocks[depth - 1]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+// Helper for transfering multiple std::vector<T> onto device with a single
+// page-locked cudaMemcpyAsync.
+struct VecPacker {
+  template <typename T>
+  // Add a vector to be copied to device
+  // NOTE: VecPacker doesn't make copies of the added vectors. They have to be
+  // kept alive by the caller until .pack() is called.
+  void add(const std::vector<T>& vec) {
+    ptrs.push_back(vec.data());
+    const auto vec_bytes = sizeof(T) * vec.size();
+    const auto vec_bytes_aligned = (vec_bytes + 16 - 1) / 16 * 16;
```

Comment:
Say I add a `std::vector<uint8_t>` of size 7, followed by a `std::vector<int64_t>`. Without alignment, the corresponding dev ptr for the `std::vector<int64_t>` would be `[allocator_alignment] + 7`, which is not aligned with `int64_t` and would lead to unaligned address error.

16 (128-bit) is just something that would cover any T here (I should probably add a `static_assert` here)

Reply:
I think you could also use the c10 round_up  (in the same file that @drisspg linked above) here?

Reply:
Yeah that would be great to mention that c10::complex<double> is the reason why we need 16 here and a static assert that no local type is bigger than that would be great.